### PR TITLE
DDF 1937 Update PreIngest Registry Creation Process to add External Identifier

### PR DESCRIPTION
**Pull request by mcalcote**
#### What does this PR do?
A new PreIngest Plugin is needed to add a generated unique external identifier ID to a Registry Metacard during the Creation Processes. If the metacard has the Registry tag, this plugin will un-marshall the metadata add the external identifier ID and re-marshall it before it is added to the catalog. 
Additionally, the registry transformer will need to set the metacard ID attribute to the external identifier ID if the input is xml.
Finally, add registry-id field to RegistryObjectMetacardType and populate it in the registry transformer with the top level id that is currently being used for the metacard id. The metacard id should no longer be populated in the registry transformer.
#### Who is reviewing it?
@clockard @vinamartin @brianfelix @gordocanchola @oscaritoro @bcwaters 
#### How should this be tested?
Run unit test, build DDF
#### Any background context you want to provide?
Builds on recent Registry changes

I will squash commits after initial feedback.

#### Checklist:
- [ ] Documentation Updated
- [x] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

### CODE REVIEW

File: catalog/registry/catalog-registry-cswinputtransformer/src/main/java/ddf/catalog/registry/converter/RegistryPackageConverter.java [hunk 1]
```diff
@@ -355,6 +356,13 @@ private static void parseNodeExtrinsicObject(RegistryObjectType registryObject,
             setAttributeFromMap(RegistryObjectMetacardType.SECURITY_LEVEL, slotMap, metacard);
         }
 
+        if (registryObject.isSetId()) {
+            String id = registryObject.getId();
+            if (StringUtils.isNotBlank(id)) {
+                metacard.setAttribute(RegistryObjectMetacardType.REGISTRY_ID, id);
```
Comment (mcalcote) [inline catalog/registry/catalog-registry-cswinputtransformer/src/main/java/ddf/catalog/registry/converter/RegistryPackageConverter.java]: Move to top level

**Comment by vinamartin** [comment]
Compilation error on bamboo.

File: catalog/registry/catalog-registry-cswinputtransformer/src/main/java/ddf/catalog/registry/converter/RegistryPackageConverter.java [hunk 2]
```diff
@@ -410,6 +419,11 @@ private static void parseTopLevel(RegistryObjectType registryObject,
             setMetacardStringAttribute(registryObject.getVersionInfo()
                     .getVersionName(), Metacard.CONTENT_TYPE_VERSION, metacard);
         }
+
+        if (registryObject.isSetExternalIdentifier()) {
+            ExternalIdentifierType extId = (registryObject.getExternalIdentifier()).get(0);
+            metacard.setId(extId.getId());
+        }
```
Comment (mcalcote) [inline catalog/registry/catalog-registry-cswinputtransformer/src/main/java/ddf/catalog/registry/converter/RegistryPackageConverter.java]: explicitly get local id extId, also get extId.getValue()

File: catalog/registry/catalog-registry-cswinputtransformer/src/main/java/ddf/catalog/registry/transformer/RegistryTransformer.java [hunk 3]
```diff
@@ -92,8 +94,12 @@ public Metacard transform(InputStream inputStream, String id)
             if (metacard == null) {
                 throw new CatalogTransformerException(
                         "Unable to transform from CSW RIM Service Record to Metacard.");
-            } else if (StringUtils.isNotEmpty(id)) {
-                metacard.setAttribute(Metacard.ID, id);
+            }
+
+            if (StringUtils.isNotEmpty(id)) {
+                metacard.setAttribute(RegistryObjectMetacardType.REGISTRY_ID, id);
```
Comment (mcalcote) [inline catalog/registry/catalog-registry-cswinputtransformer/src/main/java/ddf/catalog/registry/transformer/RegistryTransformer.java]: Should be Metacard Id, not RegistryID

File: catalog/registry/catalog-registry-cswinputtransformer/src/main/java/ddf/catalog/registry/transformer/RegistryTransformer.java [hunk 4]
```diff
@@ -144,6 +150,13 @@ private Metacard unmarshal(InputStream xmlStream)
                         (RegistryMetacardImpl) RegistryPackageConverter.getRegistryObjectMetacard(
                                 registryObjectType);
 
+                List<ExternalIdentifierType> extIds = registryObjectType.getExternalIdentifier();
+
+                if (extIds.size() > 0){
+                    String extId = extIds.get(0).getRegistryObject();
```
Comment (mcalcote) [inline catalog/registry/catalog-registry-cswinputtransformer/src/main/java/ddf/catalog/registry/transformer/RegistryTransformer.java]: Remove this bit

File: catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    private Parser parser;
+
+    private ParserConfigurator configurator;
+
+    @Override
+    public CreateRequest process(CreateRequest input)
+            throws PluginExecutionException, StopProcessingException {
+        List<Metacard> metacards = input.getMetacards();
+        Map<String, Serializable> properties = input.getProperties();
+        List<Metacard> updatedMetacards = new ArrayList<>();
```
Comment (mcalcote) [inline catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java]: getDestinations and pass into constructor below

File: catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+    private Parser parser;
+
+    private ParserConfigurator configurator;
+
+    @Override
+    public CreateRequest process(CreateRequest input)
+            throws PluginExecutionException, StopProcessingException {
+        List<Metacard> metacards = input.getMetacards();
+        Map<String, Serializable> properties = input.getProperties();
```
Comment (mcalcote) [inline catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java]: getDestinations and pass into constructor below

File: catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        List<Metacard> updatedMetacards = new ArrayList<>();
+
+        for (Metacard metacard : metacards) {
+            Set<String> metacardTags = metacard.getTags();
+            if (metacardTags.contains(RegistryConstants.REGISTRY_TAG)) {
+                updatedMetacards.add(setMetacardExtID((RegistryMetacardImpl) metacard));
```
Comment (mcalcote) [inline catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java]: Check that I need this cast

File: catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+
+                    //check if external ids are already present
+                    if (registryObjectType.isSetExternalIdentifier()) {
+                        List<ExternalIdentifierType> currentExtIdList =
+                                registryObjectType.getExternalIdentifier();
+
+                        for (ExternalIdentifierType extId : currentExtIdList) {
+                            extId.setRegistryObject(registryID);
+                            if (extId.getId()
+                                    .equals("urn:mcard:local-id")) {
```
Comment (mcalcote) [inline catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java]: use RegistryConstants

File: catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+                                //update local id
+                                extId.setValue(metacardID);
+                            }
+                            extIdList.add(extId);
+                        }
+
+                    } else {
+                        //create both ids
+                        extIdList = new ArrayList<>();
```
Comment (mcalcote) [inline catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java]: size 2

File: catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+
+                        ExternalIdentifierType localExtId = new ExternalIdentifierType();
+                        localExtId.setId("urn:mcard:local-id");
```
Comment (mcalcote) [inline catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java]: RegistryConstants

File: catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+                        localExtId.setRegistryObject(registryID);
+                        localExtId.setIdentificationScheme("MetacardId");
```
Comment (mcalcote) [inline catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java]: registryConstants (and below)

File: catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+
+                        ExternalIdentifierType originExtId = new ExternalIdentifierType();
+                        originExtId.setId("urn:mcard:origin-id");
+                        originExtId.setRegistryObject(registryID);
+                        originExtId.setIdentificationScheme("MetacardId");
+                        originExtId.setValue(metacardID);
+
+                        extIdList.add(localExtId);
+                        extIdList.add(originExtId);
+
```
Comment (mcalcote) [inline catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java]: Refactor for non-relevant External Ids

File: catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+                            .openStream());
+                    metacard.setAttribute(Metacard.METADATA, xml);
+                }
+            }
+
+        } catch (ParserException | IOException e) {
+            throw new StopProcessingException(
+                    "Unable to access Registry Metadata. Parser exception caught");
+        }
+        return metacard;
```
Comment (mcalcote) [inline catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java]: make changes in place

File: catalog/registry/catalog-registry-identificationplugin/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+ *
+ **/
+-->
+<blueprint xmlns="http://www.osgi.org/xmlns/blueprint/v1.0.0">
+
+    <bean id="identificationPlugin" class="ddf.catalog.registry.identification"/>
+
+    <service ref="identificationPlugin" interface="ddf.catalog.plugin.PreIngestPlugin" ranking="0"/>
+
+</blueprint>
```
Comment (mcalcote) [inline catalog/registry/catalog-registry-identificationplugin/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: add newline

File: catalog/registry/catalog-registry-identificationplugin/src/test/java/ddf/catalog/registry/identification/IdentificationPluginTest.java [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+        CreateRequest result = idp.process(new CreateRequestImpl(sampleData));
+
+        Metacard testMetacard = result.getMetacards()
+                .get(0);
+
+        String metadata = testMetacard.getMetadata();
+
+        InputStream inputStream = new ByteArrayInputStream(metadata.getBytes(Charsets.UTF_8));
+
+        try {
```
Comment (mcalcote) [inline catalog/registry/catalog-registry-identificationplugin/src/test/java/ddf/catalog/registry/identification/IdentificationPluginTest.java]: overkill here

File: catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+        String metadata = metacard.getMetadata();
+        String registryID = metacard.getAttribute(RegistryObjectMetacardType.REGISTRY_ID)
+                .getValue()
+                .toString();
+
+        InputStream inputStream = new ByteArrayInputStream(metadata.getBytes(Charsets.UTF_8));
+        FileBackedOutputStream outputStream = new FileBackedOutputStream(1000000);
+
+        try {
+            JAXBElement<RegistryObjectType> registryObjectTypeJAXBElement = parser.unmarshal(
```
Comment (bcwaters) [inline catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java]: Is it possible that parser is null here?

Comment (mcalcote) [inline catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java]: yes, nice catch

Comment (coyotesqrl) [inline catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java]: I'm on my phone so may be wrong, but isn't this Parser set through blueprint injection? So for it to be null you'd have blown up initializing this?

Comment (vinamartin) [inline catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java]: Could you tell me where you're setting it @mcalcote ? I'm not seeing it.

File: catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+                }
+            }
+
+        } catch (ParserException | IOException e) {
+            throw new StopProcessingException(
+                    "Unable to access Registry Metadata. Parser exception caught");
+        }
+    }
+
+    public void setParser(Parser parser) {
```
Comment (coyotesqrl) [inline catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java]: The problem seems to be that your blueprint isn't injecting your parser.

Comment (mcalcote) [inline catalog/registry/catalog-registry-identificationplugin/src/main/java/ddf/catalog/registry/identification/IdentificationPlugin.java]: Updated

**Comment by clockard** [comment]
LGTM :shipit:

**Comment by gordocanchola** [comment]
:+1:

**Pull request merged**